### PR TITLE
Enhancement issue2141

### PR DESCRIPTION
**Pull request by kg8z**
Issue: #2141

## PR Type
What kind of change does this PR introduce?

Feature

## What is the current behavior?

Filter options in the payment list for accounts

## What is the new behavior?
Added a new option to filter by payment type

## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] Tested code on Windows
- [x] Tested code on Android
- [ ] Tested code on iOS
- [ ] Tests for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Contains **NO** breaking changes

Still working on getting iOS testing working in my dev environment. Unable to validate iOS functionality.

## Other information

Note: I added a new Strings resource named AllLabel. Only the English language value has been added to the resource list.

### CODE REVIEW

File: Src/MoneyFox.Core/Queries/Payments/GetPaymentsForAccountId/GetPaymentsForAccountIdQuery.cs [hunk 1]
```diff
@@ -32,6 +34,8 @@ public class GetPaymentsForAccountIdQuery : IRequest<List<Payment>>
 
         public DateTime TimeRangeEnd { get; }
 
+        public int PaymentTypeFilter { get; set; }
```
Comment (NPadrutt) [inline Src/MoneyFox.Core/Queries/Payments/GetPaymentsForAccountId/GetPaymentsForAccountIdQuery.cs:37]: This shouldn't have a setter, since they're only supposed to be set via the constructor. The two existing of me are wrong as well so feel free to change those too :)

Comment (NPadrutt) [reply Src/MoneyFox.Core/Queries/Payments/GetPaymentsForAccountId/GetPaymentsForAccountIdQuery.cs:37]: Also I would use an enum here. Int's are much harder to understand what is meant. You could create a PaymentTypeFilter Enum with the default being "All".

File: Src/MoneyFox.Core/Queries/Payments/GetPaymentsForAccountId/GetPaymentsForAccountIdQuery.cs [hunk 2]
```diff
@@ -66,6 +70,11 @@ public Handler(IContextAdapter contextAdapter)
                     paymentQuery = paymentQuery.AreRecurring();
                 }
 
+                if (request.PaymentTypeFilter != -1)
+                {
+                    paymentQuery = paymentQuery.IsPaymentType((PaymentType)request.PaymentTypeFilter);
```
Comment (NPadrutt) [inline Src/MoneyFox.Core/Queries/Payments/GetPaymentsForAccountId/GetPaymentsForAccountIdQuery.cs:75]: This might be worth a new unit test in the GetPaymentsForAccountIdQueryTests :)

File: Src/MoneyFox.Core/(Pending)/Common/QueryObjects/PaymentQueryExtensions.cs [hunk 3]
```diff
@@ -25,6 +25,14 @@ public static IQueryable<Payment> IsExpense(this IQueryable<Payment> query)
         public static IQueryable<Payment> IsIncome(this IQueryable<Payment> query)
             => query.Where(payment => payment.Type == PaymentType.Income);
 
+        /// <summary>
+        ///     Adds a filter to a query for payments of a specified payment type
+        /// </summary>
+        /// <param name="query">Existing query.</param>
+        /// <returns>Query filtered for the specified payment type.</returns>
+        public static IQueryable<Payment> IsPaymentType(this IQueryable<Payment> query, PaymentType paymentType)
```
Comment (NPadrutt) [inline Src/MoneyFox.Core/(Pending)/Common/QueryObjects/PaymentQueryExtensions.cs:33]: That would be rather easy testable with a unit test. There are existing ones in `PaymentQueryExtensionTests`

File: Src/MoneyFox.Win/MoneyFox.Win/Converter/PaymentTypeToIntConverter.cs [hunk 4]
```diff
@@ -0,0 +1,31 @@
+﻿namespace MoneyFox.Win.Converter;
+
+using Core.Aggregates.Payments;
+using Core.Resources;
+using Microsoft.UI.Xaml.Data;
+using System;
+
+public class PaymentTypeToIntConverter : IValueConverter
+{
+    public object Convert(object value, Type targetType, object parameter, string language)
```
Comment (NPadrutt) [inline Src/MoneyFox.Win/MoneyFox.Win/Converter/PaymentTypeToIntConverter.cs:10]: This is also easy testable with a unit Test. The Convert Back Method for me doens't have to have one, when we don't use that method. But the convert method itself would be nice to be covered.

File: Src/MoneyFox/Converter/PaymentTypeIntConverter.cs [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+{
+    using Core.Aggregates.Payments;
+    using Core.Resources;
+    using System;
+    using System.Globalization;
+    using Xamarin.Forms;
+
+    public class PaymentTypeIntConverter : IValueConverter
+    {
+        public object Convert(object value, Type targetType, object parameter, CultureInfo culture)
```
Comment (NPadrutt) [inline Src/MoneyFox/Converter/PaymentTypeIntConverter.cs:11]: Same here regarding unit test :)

**Review by NPadrutt — CHANGES_REQUESTED** [review]
Thanks for your PR @kg8z 
I left you some comments. Let me know if something is unclear :)

File: Src/MoneyFox/Converter/PaymentTypeIntConverter.cs (hunk 5 above)
Comment (kg8z) [reply Src/MoneyFox/Converter/PaymentTypeIntConverter.cs:11]: Thanks for the feedback. It all makes sense. I was conflicted on how best to approach the enum. My original thought was to use a nullable variable with the existing enum but the binding wasn’t happy with null values. I considered modifying the existing enum or adding a new one but didn’t love that either, so I took the easy route and just used casting with an int. But you’re right, it makes the code more difficult to understand. I’ll create a new enum as you suggested. This should all be fairly easy to implement so I should have a PR ready in a day or two.

Comment (NPadrutt) [reply Src/MoneyFox/Converter/PaymentTypeIntConverter.cs:11]: Sounds good!
Just to be sure: you don't have to create a new PR. Just push the changes to this branch and the PR is updated automatically.

Comment (NPadrutt) [reply Src/MoneyFox/Converter/PaymentTypeIntConverter.cs:11]: Maybe to add on this: it makes sense to not use an existing enum as nullable as well as not to extend an existing one. Especially since those have a different purpose (defining a payment type vs defining a filter).
A int is an easy an cheap solution. But just creating a new enum is also quite cheap and has in my opinion some benefits over an int :)

Comment (kg8z) [reply Src/MoneyFox/Converter/PaymentTypeIntConverter.cs:11]: Sounds good to me. I'll switch from int to a new enum and push updates to the PR. Thanks!

File: Src/MoneyFox.Core/Queries/Payments/GetPaymentsForAccountId/GetPaymentsForAccountIdQuery.cs (hunk 1 above)
Comment (kg8z) [reply Src/MoneyFox.Core/Queries/Payments/GetPaymentsForAccountId/GetPaymentsForAccountIdQuery.cs:37]: Got it. I'll remove all of the setters and switch from int to enum

File: Src/MoneyFox.Core/(Pending)/Common/QueryObjects/PaymentQueryExtensions.cs (hunk 3 above)
Comment (kg8z) [reply Src/MoneyFox.Core/(Pending)/Common/QueryObjects/PaymentQueryExtensions.cs:33]: Definitely! I'll add a unit test.

File: Src/MoneyFox.Core/Queries/Payments/GetPaymentsForAccountId/GetPaymentsForAccountIdQuery.cs (hunk 2 above)
Comment (kg8z) [reply Src/MoneyFox.Core/Queries/Payments/GetPaymentsForAccountId/GetPaymentsForAccountIdQuery.cs:75]: Definitely! I'll add a unit test.

File: Src/MoneyFox.Win/MoneyFox.Win/Converter/PaymentTypeToIntConverter.cs (hunk 4 above)
Comment (kg8z) [reply Src/MoneyFox.Win/MoneyFox.Win/Converter/PaymentTypeToIntConverter.cs:10]: I deleted this class when I switched from using int to enum for the Payment Type filter. However, it was replaced with a new class called PaymentTypeFilterConverter. All of the Converter classes are in the MoneyFox.Win project which the MoneyFox.Core.Tests project does not reference. I can add a reference to the project and create a unit test if you think it's valuable, but this would suggest that none of the unit tests cover functionality in MoneyFox.Win. Am I missing something?

**Comment by kg8z** [comment]
@NPadrutt  - I've implemented all of the changes except the unit test for the additional Converter class. They should be pushed and read for review. Thanks.

File: Src/MoneyFox.Win/MoneyFox.Win/Converter/PaymentTypeToIntConverter.cs (hunk 4 above)
Comment (NPadrutt) [reply Src/MoneyFox.Win/MoneyFox.Win/Converter/PaymentTypeToIntConverter.cs]: aah, shoot. you are right. for the win project we currently do not have a test project, since with UWP it didn't run on the CI/CD pipeline. Actually have to check again with WinUI.

**Comment by NPadrutt** [comment]
@kg8z  nice! I think as soon as you fixed the merge conflict in the Strings.resx we're ready to go :)

as a tipp: the resx is a xml file. If you right click it and open it in the xml editor you can easily see the git marker.

**Review by NPadrutt — APPROVED** [review]
